examples/Workflow/WorkflowWebApp/Workflows/OrderProcessingWorkflow.cs

```diff
@@ -0,0 +1,44 @@
+﻿namespace WorkflowWebApp.Workflows
+{
+    using System.Threading.Tasks;
+    using Dapr.Workflow;
+    using WorkflowWebApp.Activities;
+
+    record OrderPayload(string Name, double TotalCost, int Quantity = 1);
+    record OrderResult(bool Processed);
+
+    class OrderProcessingWorkflow : Workflow<OrderPayload, OrderResult>
+    {
+        public override async Task<OrderResult> RunAsync(WorkflowContext context, OrderPayload order)
+        {
+            string orderId = context.InstanceId;
+
+            await context.CallActivityAsync(
+                nameof(NotifyActivity),
+                new Notification($"Received order {orderId} for {order.Name} at {order.TotalCost:c}"));
```

Comment:
We could potentially do something like a mediator pattern, but the direction I'd rather pursue is to use C# source generators to dynamically generate type-safe wrappers for making these calls. We actually already support it in the Durable Task .NET SDK. However, I want to come up with a plan for how we can support this in Dapr Workflow too (this is related to my comment about whether we really want this SDK to hide the implementation details, because it means there are certain tooling investments that can't be shared).  For this first release, I suggest we start with this and then revisit for the next big release.

Reply:
Cool. I was wondering if there was a way to do define an extension method that would allow for something like this:

````
 await context.CallActivityAsync<NotifyActivity>(
                new Notification($"Received order {orderId} for {order.Name} at {order.TotalCost:c}"));
```` 

and inside it we would only have the `nameof(T)` thing but I don't know if that is even possible in C#. From what I learning it _might_.

I think it would improve legibility but I guess that's debatable. And this extension method would probably need a better name than just `CallActivityAsync`

Reply:
You can certainly replace `nameof(NotifyActivity)` with `<NotifyActivity>(...)` but I don't think it creates enough of a better experience to justify creating this new overload. It would eliminate the problem of specifying the name of something completely wrong at compile time, but it doesn't help with the input data type validation at compile time. We could write a bunch of reflection code to solve that at runtime, but I think a better solution would be to leverage source generation in a way that solves both the name and input type problem at compile time.